add hex/7seg widget

charlesdaniels opened the pull request:
This PR adds a simple skeuomorphic "7 segment" or "hex display" widget. Screenshots are included in the README.md for the widget. It didn't seem like we had an established way to demo these widgets, so I created a demo program under `cmd/`.

andydotxyz commented on widget/hexwidget/hexwidget.go:
```diff
@@ -0,0 +1,214 @@
+
+	pos := image.Pt(0, 0)
+
+	// pt0Center := image.Pt(pos.X+int(hexWidth/2.0+hexOffset), pos.Y+int((hexHeight-2.0*hexSegmentVLength)/2))
```
Looks like old code

charlesdaniels replied:
df17679c1ae6d6f2f3ad755f0ab930a11340f923

andydotxyz commented on widget/hexwidget/hexwidget.go:
```diff
@@ -0,0 +1,214 @@
+	pt34 := image.Pt(int(float32(pt3Center.X)-(hexSegmentHLength/2)), pt3Center.Y)
+	pt32 := image.Pt(int(float32(pt3Center.X)+(hexSegmentHLength/2)), pt3Center.Y)
+
+	// segment 0
```
I think either this comment is not needed, or the `seg0` should be `segment0` if you worry it's not clear

charlesdaniels replied:
df17679c1ae6d6f2f3ad755f0ab930a11340f923

andydotxyz commented on widget/hexwidget/hexwidget.go:
```diff
@@ -0,0 +1,214 @@
+func (h *HexWidget) Set(val uint) {
+	val = val % 16
+
+	if val == 0 {
```
Go prefers `switch` over lots of if/else

charlesdaniels replied:
4f2bad7055186ccbed2afbdedab52af2e729a95c

I actually decided to make this a lookup table. There is no reason to re-compute these values at runtime, and indexing into a slice like this should be O(1) - no need for any comparisons at runtime at all.

andydotxyz commented on widget/hexwidget/hexwidget.go:
```diff
@@ -0,0 +1,214 @@
+// Set updates the hex widget to show a specific number between 0 and 15, which
+// will be rendered in hexadecimal in 0...f. If the number is greater than 15,
+// it will be modulo-ed by 16.
+func (h *HexWidget) Set(val uint) {
```
Random thought, would it be nicer to use `rune` OR would you consider consts for the `a`, rather than developer passing `10`?

charlesdaniels replied:
I don't think so. It can't really handle arbitrary runes. Also, Go allows for hexadecimal values. You can give run `h.Set(0xa)` and it will do the right thing. This way makes it easy to slice apart a bigger number with masks and set each corresponding segment individually. Using rune would make it require string processing, and consts would mean the caller would have to look stuff up.

andydotxyz commented on widget/hexwidget/hexwidget.go:
```diff
@@ -0,0 +1,214 @@
+
+// UpdateSegments changes the state of the segments and causes the widget to
+// refresh so the changes are visible to the user.
+func (h *HexWidget) UpdateSegments(segments uint8) {
```
Should this be public?

Another random thought, would it be clearer if this took a `...uint`, then did the bit management inside this method? (maybe less of an issue if this is private)

charlesdaniels replied:
df17679c1ae6d6f2f3ad755f0ab930a11340f923

charlesdaniels replied:
I have decided that this should be public, because it addresses your [other comment](https://github.com/fyne-io/fyne-x/pull/7#issuecomment-787878110) and allows the user finer-grained control if they want it.

I'm not sold on using a list of bool/uint, because then storing 1 hex state requires using a slice, which adds a lot more overhead. Additionally, users on the happy path can use the `Set()` function.

andydotxyz reviewed (changes requested):
This is a cool widget.
Some random thoughts inline, though you'd need to address the build failures as well.

charlesdaniels commented:
I am working on ironing out some of the build failures, which I will hopefully finish mopping up tonight. I also addressed all the code review comments, AFAIK.

Aside from this, I realized there was no good reason not to expose the variables to change the height/color/slant, so I exposed those in the public API, and also added some sliders/color pickers in the demo.

andydotxyz commented:
I think the slant calculations may not be working quite right, as it goes out of the bounds
<img width="728" alt="Screenshot 2021-03-01 at 11 13 41" src="https://user-images.githubusercontent.com/294436/109489724-32675480-7a7f-11eb-946c-280628c702c9.png">

andydotxyz commented on widget/hexwidget/README.md:
```diff
@@ -0,0 +1,18 @@
+## Future Work
+
+* Support scaling the size of the hex widget.
+* Support changing the hex widget color.
```
It looks like this is done

charlesdaniels replied:
bcf4e7c

andydotxyz commented on widget/hexwidget/hexwidget.go:
```diff
@@ -0,0 +1,240 @@
+func (h *hexRenderer) ApplyTheme() {
+}
+
+func (h *hexRenderer) BackgroundColor() color.Color {
```
This is not needed any more

charlesdaniels replied:
bcf4e7c

andydotxyz commented on widget/hexwidget/hexwidget.go:
```diff
@@ -0,0 +1,240 @@
+	h.segmentObjects[5].(*canvas.Line).StrokeWidth = float32(hexSegmentWidth / 2)
+	h.segmentObjects[6].(*canvas.Line).StrokeWidth = float32(hexSegmentWidth / 2)
+
+	pos := image.Pt(0, 0)
```
Using fyne.Position would be better as you would not lose precision

charlesdaniels replied:
Done in 541512d. 

This code was ported from a different toolkit, hence why there are some non-idiomatic things.

andydotxyz commented on widget/hexwidget/hexwidget.go:
```diff
@@ -0,0 +1,240 @@
+	h.Refresh()
+}
+
+func setLineEndpoints(l *canvas.Line, pt1, pt2 image.Point) {
```
I'd recommend `fyne.Position` so that it is more precisely placed

charlesdaniels replied:
541512d

andydotxyz commented on widget/hexwidget/hexwidget.go:
```diff
@@ -0,0 +1,240 @@
+	(1<<1 | (1 << 2) | (1 << 3)),
+}
+
+// size in pixels of the hex widget
```
These are not pixel sizes, the Fyne coordinate system is device independent

charlesdaniels replied:
1cb3743

andydotxyz commented on widget/hexwidget/hexwidget.go:
```diff
@@ -0,0 +1,240 @@
+	widget.BaseWidget
+	segments uint8
+
+	// size in "pixels" of the hex widget
```
Not pixels, sorry.
Also, why not just use `fyne.Size` for a single field?

charlesdaniels replied:
2f597c1

andydotxyz commented on widget/hexwidget/hexwidget.go:
```diff
@@ -0,0 +1,240 @@
+}
+
+// SetSlant changes the amount of "slant" i nthe hex widgets. The topmost
+// segment is offset by slant many virtual pixels to the right. A value of 0
```
Not sure that "virtual pixels" means much, sounds like pixels again.

andydotxyz replied:
Also, for clarity, is it worth stating (if this is true?) that setting slant = height will give 45degree slope?

charlesdaniels replied:
AFAIK, this is correct.

bcf4e7c

andydotxyz commented on widget/hexwidget/hexwidget.go:
```diff
@@ -0,0 +1,240 @@
+	h.Refresh()
+}
+
+// SetSlant changes the amount of "slant" i nthe hex widgets. The topmost
```
```suggestion
// SetSlant changes the amount of "slant" in the hex widgets. The topmost
```

charlesdaniels replied:
1cb3743

andydotxyz reviewed (changes requested):
Looks cool, but some docs and/or coordinates could be improved

andydotxyz commented:
One other random observation... The widget starts with no segments set, but this is a state it can never return to.
I'd say that either it should either start with "0" visible, or have a way to go back to "off".

charlesdaniels commented:
I believe this is once again ready for review at your leisure.

andydotxyz reviewed (approved):
Such a cool widget, thanks :)
